DOC: Use internal/intersphinx links for neps.

rossbar opened the pull request:
Inspired by #27920 .

Replace external links to NEPs with explicit targets with the appropriate sphinx (or intersphinx) roles. A very minor tweak that should improve the robustness of links moving forward.

These are just the instances I found by grepping (ignoring e.g. release notes).

rossbar committed:
DOC: Use internal/intersphinx links for neps.

[skip cirrus] [skip azp] [skip actions]

rossbar commented on doc/neps/index.rst:
```diff
@@ -6,8 +6,7 @@ This page provides an overview of development priorities for NumPy.
 to the library—in various stages of discussion or completion.
-See `NEP 0 <https://github.com/numpy/numpy/blob/main/doc/neps/nep-0000.rst>`__
-for more informations about NEPs.
+See :doc:`nep-0000` for more informations about NEPs.
```
This of course will also change the link wording to include the title of the NEP. If there's a strong preference to preserve the exact link rendering, please don't hesitate to do so (e.g. ``:doc:`NEP 0 <nep-0000>``)

mattip commented:
There are a few more, I used a different grep `$ git grep "neps/nep" | grep -v "doc/source/release"`
- [ ] `doc/C_STYLE_GUIDE.rst:"NEP 45 — C Style Guide" at https://numpy.org/neps/nep-0045-c_style_guide.html`
- [x] `doc/neps/index.rst:See NEP 0 <https://github.com/numpy/numpy/blob/main/doc/neps/nep-0000.rst>__`
- [ ] `doc/neps/nep-0021-advanced-indexing.rst:   http://www.numpy.org/neps/nep-0018-array-function-protocol.html`
- [x] `doc/neps/nep-0022-ndarray-duck-typing-overview.rst:<http://www.numpy.org/neps/nep-0021-advanced-indexing.html>`__ of`
- [ ] `doc/neps/nep-0023-backwards-compatibility.rst: NEP 19 <http://www.numpy.org/neps/nep-0019-rng-policy.html>_.`
- [ ] `doc/source/dev/index.rst:   NumPy C style guide <https://numpy.org/neps/nep-0045-c_style_guide.html>`
- [x] `doc/source/reference/simd/index.rst:.. _NEP-38: https://numpy.org/neps/nep-0038-SIMD-optimizations.html`
- [ ] `numpy/lib/mixins.py:    <https://numpy.org/neps/nep-0013-ufunc-overrides.html>_.`

charris commented:
LGTM. I wonder if `nep-0000` should have a bit more in the title?

rossbar committed:
DOC: Suggestions from review.

[skip cirrus] [skip azp] [skip actions]

rossbar commented:
Thanks @mattip , I've fixed up the stragglers.

I had originally left the one in `C_STYLE_GUIDE.rst` intentionally alone because I saw that document was referenced in `.github/CONTRIBUTING.md` which tries to render rst but doesn't have sphinx so sphinx-y directives/roles are not supported. However, upon closer inspection it doesn't look like it's `.. included` or anyting like that so it should be safe to update the link :+1: 

>  I am not sure about the last one, help(np.lib.mixins.NDArrayOperatorsMixin) will no longer display a valid link

I've updated it and tested that it renders okay in html :+1: 

> I wonder if nep-0000 should have a bit more in the title?

`nep-0000` is just the target name in the `objects.inv` file, though the target itself could be updated in the NEP rst if desired!

charris commented:
> nep-0000 is just the target name

I was suggesting renaming the file. The other nep file names have a hint of their contents.
